### ISSUE TITLE
Partial reduce and failure

### ISSUE THREAD

**Issue by Maxwell-Rosen** [ammarhakim/gkylzero#307]
There is an issue with the multi-gpu reduction operation in our code. It is most prevalent and makes an appearance in our production runs during the mirror regression tests. When run on 1 GPU, the time step is much larger than when the code is run on many GPUs, for instance 8, and the timestep becomes much smaller. This issue is reproduced in the unit test ctest_array_reduce in the "big" unit test, where the reduce doesn't seem to be finding the maximum value in the array when it is on the GPU. This seems like an issue that NVIDIA should be contacted about because we are using their functions and they aren't working as intended.

**Comment by JunoRavin** [comment]
This bug has been fixed as of: https://github.com/ammarhakim/gkylzero/commit/ce9739aa20877ba893d0b97de57fdfc7b5450edb

The problem was what we were doing before did not seem to do the *full* reduction over threads (we utilize CUB to do our reduction operations over blocks, and then it is left to us to do the final reduction over the threads). For whatever reason the operation we copied from the Nvidia example with the do while loop here: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions

did not robustly work. We have instead copied an implementation from ye olden times in CUDA: https://github.com/treecode/Bonsai/blob/master/runtime/profiling/derived_atomic_functions.h#L226

that appears to robustly work regardless of block size.
